### ISSUE TITLE
Order of logs during start/stop of container

### ISSUE THREAD

**Issue by kssxs** [petersem/monocker#101]
Hi there,

I noticed that when a container is stopped, and subsequently restarted, when Monocker reports the stop and start, it reverses the order in which it displays in the logs (and also in my Discord notification).  Essentially, it makes it looks like the container started, and then immediately stopped, but when I check Portainer, or Uptime Kuma to confirm the status of the container, everything is found to be running normally.  I'm not sure if it's related to how little time goes by in between start and stop events as sometimes the time period from when a container stops and then is restarted can be very short, but that shouldn't matter (I think).  It's been doing this since I installed it a couple of weeks ago.

![image](https://github.com/user-attachments/assets/3f2107e9-1f6e-4dd9-a336-0ea40986f1e9)

![image](https://github.com/user-attachments/assets/ccd5054b-ce2e-4180-b436-b1de2ccba84b)

It seems that Monocker is reporting the events in the incorrect order, as you would expect the last log entry to be the current state, making it look like the container was started, and then is stopped, when in fact it's running normally.  AFAIK, most logs display in order of latest entry (current state) last.

Can this be looked at, or fixed, or is there an environment variable that I need to set to fix it?

Thanks in advance!

**Comment by petersem** [comment]
@kssxs I will have a look. can you please advise what you have set for the 'period' value?

**Comment by kssxs** [comment]
@petersem I'm using 10s for the Period value.  My yaml with Discord channel info removed is below.  Hope this helps!

YAML:
=======================================
services:
  Monocker-Container-Monitor:

    cap_drop:
      - "AUDIT_CONTROL"
      - "BLOCK_SUSPEND"
      - "DAC_READ_SEARCH"
      - "IPC_LOCK"
      - "IPC_OWNER"
      - "LEASE"
      - "LINUX_IMMUTABLE"
      - "MAC_ADMIN"
      - "MAC_OVERRIDE"
      - "NET_ADMIN"
      - "NET_BROADCAST"
      - "SYSLOG"
      - "SYS_ADMIN"
      - "SYS_BOOT"
      - "SYS_MODULE"
      - "SYS_NICE"
      - "SYS_PACCT"
      - "SYS_PTRACE"
      - "SYS_RAWIO"
      - "SYS_RESOURCE"
      - "SYS_TIME"
      - "SYS_TTY_CONFIG"
      - "WAKE_ALARM"

    command:
      - "node"
      - "index.js"

    container_name: "Monocker-Container-Monitor"

    environment:
      - "PGID=100"
      - "PUID=1004"
      - "TZ=America/Vancouver"
      - "MESSAGE_PLATFORM=discord@https://discord.com/api/webhooks/randomlettersandnumbers"
      - "DISABLE_STARTUP_MSG=true"
      - "PERIOD=10"
      - "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"

    expose:
      - "8000/tcp"

    hostname: "Monocker-Container-Monitor"

    image: "petersem/monocker:latest"

    ipc: "private"

    logging:
      driver: "json-file"
      options:
        max-file: "10"
        max-size: "10m"

    mac_address: "02:42:0a:00:03:05"

    network_mode: "bridge"

    restart: "unless-stopped"

    ulimits:
      - Hard: 65535
        Name: "nofile"
        Soft: 65535

    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"

    working_dir: "/app"

version: "3.6"


=======================================

**Comment by petersem** [comment]
@kssxs I cant replicate this out of order issue. What are you doing for these messages to come out?

**Comment by kssxs** [comment]
@petersem Hey, sorry for the delay in responding.

I have a setup with both DeUnhealth (restarts unhealthy containers) and Watchtower (nightly container updates).  Here is the yaml for both:

`services:

  Docker-DeUnhealth-Restarter:

    cap_drop:
      - "AUDIT_CONTROL"
      - "BLOCK_SUSPEND"
      - "DAC_READ_SEARCH"
      - "IPC_LOCK"
      - "IPC_OWNER"
      - "LEASE"
      - "LINUX_IMMUTABLE"
      - "MAC_ADMIN"
      - "MAC_OVERRIDE"
      - "NET_ADMIN"
      - "NET_BROADCAST"
      - "SYSLOG"
      - "SYS_ADMIN"
      - "SYS_BOOT"
      - "SYS_MODULE"
      - "SYS_NICE"
      - "SYS_PACCT"
      - "SYS_PTRACE"
      - "SYS_RAWIO"
      - "SYS_RESOURCE"
      - "SYS_TIME"
      - "SYS_TTY_CONFIG"
      - "WAKE_ALARM"

    container_name: "Docker-DeUnhealth-Restarter"

    entrypoint:
      - "/app"

    environment:
      - "PGID=100"
      - "PUID=1004"
      - "TZ=America/Vancouver"
      - "DOCKER_HOST=unix:///var/run/docker.sock"
      - "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
      - "LOG_LEVEL=info"
      - "HEALTH_SERVER_ADDRESS=127.0.0.1:9999"

    hostname: "Docker-DeUnhealth-Restarter"

    image: "qmcgaw/deunhealth:latest"

    ipc: "private"

    labels:
      org.opencontainers.image.authors: "quentin.mcgaw@gmail.com"
      org.opencontainers.image.created: "2021-11-28T12:51:56Z"
      org.opencontainers.image.description: "REstart your unhealthy containers in style"
      org.opencontainers.image.documentation: "https://github.com/qdm12/deunhealth/blob/main/README.md"
      org.opencontainers.image.revision: "6f26ebf"
      org.opencontainers.image.source: "https://github.com/qdm12/deunhealth"
      org.opencontainers.image.title: "deunhealth"
      org.opencontainers.image.url: "https://github.com/qdm12/deunhealth"
      org.opencontainers.image.version: "latest"

    logging:
      driver: "json-file"
      options:
        max-file: "10"
        max-size: "10m"

    mac_address: "02:42:0a:00:03:17"

    network_mode: "bridge"

    privileged: true

    restart: "unless-stopped"

    security_opt:
      - "label=disable"

    ulimits:
      - Hard: 65535
        Name: "nofile"
        Soft: 65535

    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"

    working_dir: "/"
	
	Watchtower:

    cap_drop:
      - "AUDIT_CONTROL"
      - "BLOCK_SUSPEND"
      - "DAC_READ_SEARCH"
      - "IPC_LOCK"
      - "IPC_OWNER"
      - "LEASE"
      - "LINUX_IMMUTABLE"
      - "MAC_ADMIN"
      - "MAC_OVERRIDE"
      - "NET_ADMIN"
      - "NET_BROADCAST"
      - "SYSLOG"
      - "SYS_ADMIN"
      - "SYS_BOOT"
      - "SYS_MODULE"
      - "SYS_NICE"
      - "SYS_PACCT"
      - "SYS_PTRACE"
      - "SYS_RAWIO"
      - "SYS_RESOURCE"
      - "SYS_TIME"
      - "SYS_TTY_CONFIG"
      - "WAKE_ALARM"

    container_name: "Watchtower"

    entrypoint:
      - "/watchtower"

    environment:
      - "PGID=100"
      - "PUID=1004"
      - "TZ=America/Vancouver"
      - "WATCHTOWER_CLEANUP=true"
      - "WATCHTOWER_INCLUDE_RESTARTING=true"
      - "WATCHTOWER_ROLLING_RESTART=true"
      - "WATCHTOWER_TIMEOUT=30s"
      - "WATCHTOWER_SCHEDULE=0 0 4 * * *"
      - "WATCHTOWER_NOTIFICATIONS=email"
      - "WATCHTOWER_NOTIFICATION_EMAIL_FROM=paul.chera@gmail.com"
      - "WATCHTOWER_NOTIFICATION_EMAIL_TO=paul.chera@gmail.com"
      - "WATCHTOWER_NOTIFICATION_EMAIL_SERVER=smtp.gmail.com"
      - "WATCHTOWER_NOTIFICATION_EMAIL_SERVER_PORT=587"
      - "WATCHTOWER_NOTIFICATION_EMAIL_SERVER_USER=paul.chera@gmail.com"
      - "WATCHTOWER_NOTIFICATION_EMAIL_SERVER_PASSWORD=isswxeobxmotbstu"
      - "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
      - "WATCHTOWER_NOTIFICATION_REPORT=true"
      - "WATCHTOWER_NOTIFICATION_URL=generic://notifiarr.com/api/v1/notification/watchtower/bb8855c3-40a4-4e59-9499-441034154412?template=json&server=NAME"

    expose:
      - "8080/tcp"

    hostname: "Watchtower"

    image: "containrrr/watchtower:latest"

    ipc: "private"

    labels:
      com.centurylinklabs.watchtower: "true"
      deunhealth.restart.on.unhealthy: "true"

    logging:
      driver: "json-file"
      options:
        max-file: "10"
        max-size: "10m"

    mac_address: "02:42:0a:00:03:11"

    network_mode: "bridge"

    restart: "always"

    ulimits:
      - Hard: 65535
        Name: "nofile"
        Soft: 65535

    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"

    working_dir: "/"

version: "3.6"`

I also have a weekly NAS reboot on Thursday at 4am.  Monocker is showing the same for whether the container is restarted by DeUnhealth, or Watchtower, or the weekly NAS reboot, with the incorrect order of operations occurring.

In the grand scheme of things, everything is working correctly, but just the order of entries in the log and Discord notification is reversed.

Hope that helps you find the issue (assuming there is one).

Thanks!

Monocker just informs me when a container is restarted, but in the incorrect order based

**Comment by petersem** [comment]
@kssxs can you replicate this manually,

I have watchtower and autoheal (like deunhealth), yet don't see this.

Can you try monocker with an interval of 30 seconds?

![Screenshot_20240816_122708_Discord](https://github.com/user-attachments/assets/76e39221-1e60-437b-a2c9-b4330188f882)

**Comment by kssxs** [comment]
Ok, done, changed to 30s.  I'll monitor and let you know what happens:

![image](https://github.com/user-attachments/assets/a0c73efd-138e-4eb4-b197-ac616421031f)

**Comment by kssxs** [comment]
It's still doing it:

![image](https://github.com/user-attachments/assets/053af1fc-5a8e-41a4-a828-6ae02205b60d)

Really weird

**Comment by petersem** [comment]
@kssxs Ok, i've found the reason for this. When a container is auto upgraded by another app, a new container is created, the old one is deleted, and the new one is renamed. Depending on the timings for Monocker, this sometimes looks like a container is starting then stopping, but these are actually two seperate containers. 

If you turn on the environment option `SHA: 'true'`, then you will see the ID's change. In the case of exited containers, there either wont be an ID or it will be undefined. 

TLDR: It's all working fine, but can be confusing with output if SHA is turned off. So either turn on SHA, or use the `EXCLUDE_EXITED: 'true'` option.

**Comment by kssxs** [comment]
I'll try with the SHA option and see how it goes.  I guess factoring in the change in container ID can't be written in?  I doubt that is something quick to implement (I'm not a programmer, lol), but it's worth asking either way :) .

Thanks for explaining!
